### PR TITLE
Add interpreter for CollectiveBroadcastOp

### LINKED ISSUE #1982: Add interpreter for CollectiveBroadcastOp
Following the [testing guidelines](https://github.com/openxla/stablehlo/blob/main/docs/interpreter_checklist.md)

### PR DESCRIPTION
**Pull request by ghpvnist**
closes #1982

### CODE REVIEW

File: stablehlo/reference/Ops.cpp [hunk 1]
```diff
@@ -328,6 +328,25 @@ SmallVector<InterpreterValue> eval(Region &region,
...
+      auto replicaGroupsShape = replicaGroupsAttr.getShapedType().getShape();
+      SmallVector<SmallVector<uint32_t>> replicaGroups(replicaGroupsShape[0]);
+      auto replicaGroupsIt = replicaGroupsAttr.getValues<int64_t>().begin();
+      for (auto &replicaGroup : replicaGroups)
+        for (auto i = 0; i < replicaGroupsShape[1]; ++i, ++replicaGroupsIt)
+          replicaGroup.push_back(*replicaGroupsIt);
+
+      ChannelId channelId = 0;
+      if (auto channelHandle = collectiveBroadcastOp.getChannelHandle())
+        channelId = channelHandle->getHandle();
```
Comment (mlevesquedion) [inline stablehlo/reference/Ops.cpp:345]: Nit: the difference in names here and in the spec is a little bit confusing.  Should `getHandle()` be `getId()` ? (This is outside the scope of this CL, I'm just wondering.)

File: stablehlo/reference/Ops.cpp [hunk 2]
```diff
@@ -328,6 +328,25 @@ SmallVector<InterpreterValue> eval(Region &region,
...
+                   dyn_cast<CollectiveBroadcastOp>(op)) {
+      auto operand = scope.findTensor(collectiveBroadcastOp.getOperand());
+
+      auto replicaGroupsAttr = collectiveBroadcastOp.getReplicaGroups();
+      auto replicaGroupsShape = replicaGroupsAttr.getShapedType().getShape();
+      SmallVector<SmallVector<uint32_t>> replicaGroups(replicaGroupsShape[0]);
+      auto replicaGroupsIt = replicaGroupsAttr.getValues<int64_t>().begin();
+      for (auto &replicaGroup : replicaGroups)
+        for (auto i = 0; i < replicaGroupsShape[1]; ++i, ++replicaGroupsIt)
+          replicaGroup.push_back(*replicaGroupsIt);
```
Comment (mlevesquedion) [inline stablehlo/reference/Ops.cpp:341]: Would a call to `getReplicaGroups` work here? I'm noticing that `AllToAll` and `AllGather` also have this code.

File: stablehlo/tests/interpret/collective_broadcast.mlir [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+    %operand3 = stablehlo.constant dense<[[7, 8]]> : tensor<1x2xi64>
+    %results:4 = "interpreter.run_parallel"(%operand0, %operand1, %operand2, %operand3) {
+      programs=[[@collective_broadcast, @collective_broadcast,
+                 @collective_broadcast, @collective_broadcast]]
+    } : (tensor<1x2xi64>, tensor<1x2xi64>, tensor<1x2xi64>, tensor<1x2xi64>) ->
+        (tensor<1x2xi64>, tensor<1x2xi64>, tensor<1x2xi64>, tensor<1x2xi64>)
+    check.expect_eq_const %results#0, dense<[[1, 2]]> : tensor<1x2xi64>
+    check.expect_eq_const %results#1, dense<[[3, 4]]> : tensor<1x2xi64>
+    check.expect_eq_const %results#2, dense<[[5, 6]]> : tensor<1x2xi64>
+    check.expect_eq_const %results#3, dense<[[7, 8]]> : tensor<1x2xi64>
```
Comment (mlevesquedion) [inline stablehlo/tests/interpret/collective_broadcast.mlir:80]: I'm a bit confused as to why this is the expected result. Based on the previous two examples, I would have expected the output to be [[1, 2]], [[0, 0]], [[0, 0]], [[0, 0]]. What am I missing?

**Review by mlevesquedion — COMMENTED** [review]
Change looks good but I'm confused about how replicaGroups / processGroups work. I will need to do some more digging into those before I can properly review.

File: stablehlo/tests/interpret/collective_broadcast.mlir (hunk 3 above)
Comment (mlevesquedion) [reply stablehlo/tests/interpret/collective_broadcast.mlir:80]: Please ignore this. I didn't notice that the `programs` argument differed between tests and I didn't understand its significance.

File: stablehlo/tests/interpret/collective_broadcast.mlir [hunk 4]
```diff
@@ -0,0 +1,223 @@
+// RUN: stablehlo-translate --interpret -split-input-file %s
+
+module @cross_replica {
+  func.func public @collective_broadcast(%operand : tensor<1x2xi64>) -> tensor<1x2xi64> {
```
[the remark this answers is not in the document]
Comment (mlevesquedion) [reply stablehlo/tests/interpret/collective_broadcast.mlir:4]: According to this documentation symbols are public by default: https://mlir.llvm.org/docs/SymbolsAndSymbolTables/#symbol-visibility

Consider removing the specifier as it doesn't seem relevant for these tests.

File: stablehlo/reference/Ops.cpp [hunk 5]
```diff
@@ -1074,6 +1093,28 @@ Tensor evalClzOp(const Tensor &operand, ShapedType resultType) {
...
+    const Tensor &operand, SmallVector<SmallVector<uint32_t>> replicaGroups,
+    ChannelId channelId, Process *process) {
+  if (!process)
+    llvm::report_fatal_error(
+        "collective_broadcast is only supported when run via "
+        "interpreter.run_parallel");
+
+  ProcessGroups processGroups;
+  if (channelId <= 0) processGroups = process->crossReplica(replicaGroups);
+  if (channelId > 0) processGroups = process->crossPartition(replicaGroups);
```
Comment (mlevesquedion) [inline stablehlo/reference/Ops.cpp:1106]: Not blocking, outside the scope of this PR: I find this way of creating the process groups unintuitive. Why not use an enum to decide which communication scheme to use?

Also, when using `crossPartition` the `replicaGroups` are actually `partitionGroups`. Would `communicationGroups` be a better name?

File: stablehlo/reference/Ops.cpp [hunk 6]
```diff
@@ -1074,6 +1093,28 @@ Tensor evalClzOp(const Tensor &operand, ShapedType resultType) {
...
+
+  for (auto processGroup : processGroups)
+    if (llvm::find(processGroup, process->getId()) != processGroup.end())
+      return process->rendezvous(processGroup, channelId, operand)
+          .lookup(processGroup[0]);
```
Comment (mlevesquedion) [inline stablehlo/reference/Ops.cpp:1111]: Would `processGroups.findGroup` work here? Perhaps you can update the implementation of that to use `llvm::find`, perhaps in a separate change.

File: stablehlo/tests/interpret/collective_broadcast.mlir
Comment (mlevesquedion) [inline stablehlo/tests/interpret/collective_broadcast.mlir:1]: Nit: since the behavior of the operation depends so heavily on the way the process groups are formed, consider adding comments to each test case showing what the process groups look like.

**Review by mlevesquedion — APPROVED** [review]

File: stablehlo/reference/Ops.cpp (hunk 1 above)
Comment (ghpvnist) [reply stablehlo/reference/Ops.cpp:345]: This is one of those things that are part of the Frontend Contract workstream. We have a issue in #351 to track this, and I've added your comment.

File: stablehlo/reference/Ops.cpp (hunk 5 above)
Comment (ghpvnist) [reply stablehlo/reference/Ops.cpp:1106]: To answer the first question, it's written like that for spec parity since the spec uses them as function calls: 
https://openxla.org/stablehlo/spec#collective_broadcast
https://openxla.org/stablehlo/spec#collective_ops

File: stablehlo/reference/Ops.cpp (hunk 2 above)
Comment (ghpvnist) [reply stablehlo/reference/Ops.cpp:341]: `getReplicaGroups` returns a `DenseIntElementsAttr`, but we still need to convert it to a list of lists to pass it to the eval function.

File: stablehlo/tests/interpret/collective_broadcast.mlir
Comment (ghpvnist) [reply stablehlo/tests/interpret/collective_broadcast.mlir:1]: Hm, this is certainly good to have. I'll think about this and address them separately since this would require nontrivial amount of work.

File: stablehlo/tests/interpret/collective_broadcast.mlir (hunk 4 above)
[the remark this answers is not in the document]
Comment (ghpvnist) [reply stablehlo/tests/interpret/collective_broadcast.mlir]: Done.

File: stablehlo/reference/Ops.cpp (hunk 6 above)
Comment (ghpvnist) [reply stablehlo/reference/Ops.cpp]: Done.

File: stablehlo/reference/Ops.cpp [hunk 7]
```diff
@@ -1105,10 +1105,10 @@ Tensor evalCollectiveBroadcastOp(
   if (channelId <= 0) processGroups = process->crossReplica(replicaGroups);
   if (channelId > 0) processGroups = process->crossPartition(replicaGroups);
 
-  for (auto processGroup : processGroups)
-    if (llvm::find(processGroup, process->getId()) != processGroup.end())
-      return process->rendezvous(processGroup, channelId, operand)
-          .lookup(processGroup[0]);
+  auto processGroup = processGroups.findGroup(process->getId());
+  if (processGroup)
```
Comment (mlevesquedion) [inline stablehlo/reference/Ops.cpp:1109]: Nit: `if (auto processGroup = ...) ...` would be slightly cleaner, since `processGroup` only needs to be in scope inside the `if`.

File: stablehlo/reference/Ops.cpp (hunk 2 above)
Comment (mlevesquedion) [reply stablehlo/reference/Ops.cpp:341]: Ah good point. I notice that this code is duplicated in a few places though, consider adding a helper.

File: stablehlo/reference/Ops.cpp (hunk 7 above)
Comment (ghpvnist) [reply stablehlo/reference/Ops.cpp:1109]: Ack. This pattern exists throughout the file, and I plan to create a PR to clean them up later.
